torch/csrc/autograd/function.cpp

```diff
@@ -46,10 +46,16 @@ static void gatherFunctions(
   func->release_variables();
 
   for (auto& edge : func->next_edges()) {
-    if (edge.function.use_count() == 1) {
-      stack.emplace_back(std::move(edge.function));
+    auto & fn = edge.function;
+    // When using DataParallel, modules will be replcated multiple times, with
```

Comment:
Module replication has nothing to do with autograd graph. Every single replicated module will create an independent part of the graph, and `use_count`s should be updated accordingly.

Reply:
I can verify that in the failing case mentioned in #16532, there are 3 `shared_ptr`s pointing to the same `BroadcastBackward` function object, and the `use_count()` on all 3 `shared_ptr`s are 1. I think it starts from [here](https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/python_function.cpp#L1096) where we create `shared_ptr`s from raw pointers. It was not a problem prior to 89d56ae, because we rely on Python to determine when to delete a function. However, after 89d56ae, [C++ code](https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/function.cpp#L49) will remove function edges from autograd graph based on `shared_ptr` `use_count`.

Reply:
Functions inherit from `shared_from_this`, so we should just replace this line with `self->cdata->shared_from_this()`